File: src/main/java/graveler/collection/UniquePriorityQueue.java
package graveler.collection;

import java.util.Comparator;
import java.util.HashSet;
import java.util.PriorityQueue;

public class UniquePriorityQueue<E> {
  private HashSet<E> tracker;
  private PriorityQueue<E> queue;

  public UniquePriorityQueue(int initialCapacity, Comparator<? super E> comparator) {
    tracker = new HashSet<E>();
    queue = new PriorityQueue<E>(initialCapacity, comparator);
  }

  public E remove() {
    final E element = queue.remove();
    tracker.remove(element);

    return element;
  }

  public boolean add(final E elem) {
    if (tracker.contains(elem)) {
      return false;
    }

    tracker.add(elem);
    queue.add(elem);

    return true;
  }

  public boolean isEmpty() {
    return queue.isEmpty();
  }

  public int size() {
    return queue.size();
  }
}

File: src/main/kotlin/graveler/collection/UniquePriorityQueue.kt
package graveler.collection

import java.util.*

class UniquePriorityQueue<E>(initialCapacity: Int, comparator: Comparator<in E>) {
  val tracker = HashSet<E>()
  val queue = PriorityQueue(initialCapacity, comparator)

  fun remove(): E {
    val elem = queue.remove()
    tracker.remove(elem)

    return elem
  }

  fun add(elem: E): Boolean {
    if (tracker.contains(elem)) {
      return false
    }

    tracker.add(elem)
    queue.add(elem)

    return true
  }

  val isEmpty: Boolean
    get() = queue.isEmpty()

  val size: Int
    get() = queue.size
}
